### PR TITLE
[TableGen] Allow empty terminator in SequenceToOffsetTable

### PR DESCRIPTION
**Pull request by s-barannikov**
Some clients do not want to emit a terminator after each sub-sequence (they have other means of determining the length of sub-sequences).

This moves `Term` argument from `emit` method to the constructor and makes it optional. It couldn't be made optional while still on the `emit` method because if the terminator wasn't specified, it has to be taken into account in `layout` method as well.

The fact that `layout` method was called is now recorded in a dedicated member variable, `IsLaidOut`. `Entries != 0` can no longer be used to reliably check if `layout` method was called because it may be zero for a different reason: the terminator wasn't specified and all added sequences (if any) were empty.

This reduces the size of `*LaneMaskLists` and `*SubRegIdxLists` a bit and resolves the removed TODO.

### CODE REVIEW

File: llvm/utils/TableGen/Basic/SequenceToOffsetTable.h [hunk 1]
```diff
@@ -97,25 +103,27 @@ class SequenceToOffsetTable {
...
+    assert(!IsLaidOut && "Can only call layout() once");
+    IsLaidOut = true;
+
     // Lay out the table in Seqs iteration order.
     for (typename SeqMap::iterator I = Seqs.begin(), E = Seqs.end(); I != E;
          ++I) {
       I->second = Entries;
       // Include space for a terminator.
-      Entries += I->first.size() + 1;
+      Entries += I->first.size() + Terminator.has_value();
```
Comment (jurahul) [inline llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:120]: nit: Can we just do:

```
if (Terminator.has_value())
  Entries += Seqs.size();
```

after the loop?

File: llvm/utils/TableGen/Basic/SequenceToOffsetTable.h [hunk 2]
```diff
@@ -127,10 +135,10 @@ class SequenceToOffsetTable {
   /// `\0`. Falls back to emitting a comma-separated integer list if
   /// `EmitLongStrLiterals` is false
   void emitStringLiteralDef(raw_ostream &OS, const Twine &Decl) const {
-    assert(Entries && "Call layout() before emitStringLiteralDef()");
+    assert(IsLaidOut && "Call layout() before emitStringLiteralDef()");
     if (!EmitLongStrLiterals) {
       OS << Decl << " = {\n";
-      emit(OS, printChar, "0");
+      emit(OS, printChar);
```
Comment (jurahul) [inline llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:141]: Do we need to print the terminator here?

**Comment by s-barannikov** [comment]
There is a bunch of warnings I need to deal with: `warning: zero size arrays are an extension [-Wzero-length-array]`
that turn into an error when building with MSVC: `error C2466: cannot allocate an array of constant size 0`.

File: llvm/utils/TableGen/Basic/SequenceToOffsetTable.h (hunk 2 above)
Comment (s-barannikov) [reply llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:141]: It is printed by `emit` after each subsequence, assuming the terminator was provided.
The line below adds an extra '0' that I believe it should not do (and I think I ran into this before), but this is a pre-existing behavior and may be related to the issue I'm observing (see my comment below).

Comment (s-barannikov) [reply llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:141]: The extra '\0' printing was added in https://reviews.llvm.org/D126136 and looks intentional. This way the generated data is the same regardless of EmitLongStrLiterals. If the option is enabled, the whole data is printed as a single string literal, and this string literal has an implicit terminating '\0' according to C. That change made the other case emit this extra '\0'.

It doesn't appear related to the issue with zero-sized arrays.

File: llvm/utils/TableGen/Basic/SequenceToOffsetTable.h (hunk 1 above)
Comment (s-barannikov) [reply llvm/utils/TableGen/Basic/SequenceToOffsetTable.h]: Done

**Comment by s-barannikov** [comment]
> There is a bunch of warnings I need to deal with: `warning: zero size arrays are an extension [-Wzero-length-array]` that turn into an error when building with MSVC: `error C2466: cannot allocate an array of constant size 0`.
> 
> ```
>     84  extern const uint16_t XCoreSubRegIdxLists[] = {
>     85    /* 0 */ 
>     86  };
> ```

I couldn't think of anything better than to print a dummy default-constructed element if the array would otherwise be empty.

File: llvm/utils/TableGen/Basic/SequenceToOffsetTable.h (hunk 1 above)
Comment (s-barannikov) [reply llvm/utils/TableGen/Basic/SequenceToOffsetTable.h]: Unfortunately, that didn't work out. `I->second` above needs the adjusted value.

Comment (jurahul) [reply llvm/utils/TableGen/Basic/SequenceToOffsetTable.h:120]: I see that, thanks for trying it out.

**Review by jurahul — APPROVED** [review]

**Pull request merged**